### DISCUSSION TITLE
The middleware's catch block does not get triggered when errors occur in a controller that extends a `ServiceEndpoint`.

### DISCUSSION THREAD

**Discussion by EdwinThomas11** [serenity-is/Serenity#7027] [Help] [answered]
public async Task Invoke(HttpContext context)
        {
            try
            {
                await _next(context);
            }
            catch (Exception ex)
            {
                context.Response.StatusCode = 400;
                await context.Response.WriteAsync("Error occured while porcessing the request");
            }             
        }
   
   If "ServiceEndpoint" is removed from the controller then the catch block is getting triggered.

**Comment by EdwinThomas11** [comment]
@volkanceylan @VictorTomaili please help here.

**Reply by VictorTomaili** [reply]
Check this 
https://learn.microsoft.com/en-us/aspnet/core/fundamentals/middleware/?view=aspnetcore-8.0#middleware-order

**Reply by EdwinThomas11** [reply]
@VictorTomaili The middleware functions as intended when the ServiceEndpoint is excluded from the endpoints. I presume that the ServiceEndpoint internally manages errors, preventing the middleware catch from being triggered when an error occurs within the method. 

Nevertheless, I have addressed this issue by overriding the ServiceEndpoint methods. 

I am interested in understanding how ServiceEndpoint prevents the catch block in the middleware from being executed.

**Reply by VictorTomaili** [reply, marked as answer]
The `ServiceEndpoint` has an `HandleServiceExceptionAttribute` applied to it. This attribute inherits from `ExceptionFilterAttribute`. The filter catches exceptions in the inner scope. This means that if the controller scope has such an attribute or filter that handles the exception, the middleware or other ways of handling exceptions provided by ASP.NET may not be used. The filter takes the exception, deals with it according to the logic in the filter, and may stop it from going up to the higher levels of the application where middleware could handle it otherwise.

**Reply by EdwinThomas11** [reply]
Understood, @VictorTomaili . Appreciate the prompt reply.
